external_services/lti.py

```diff
@@ -93,9 +93,8 @@ def __init__(self, service, user, instance, request, title, context_id=None, lin
     def user_info(self, course_instance, user):
         if self.service.is_anonymous:
             # Anonymize user information
-            enrollment = Enrollment.objects.filter(course_instance=course_instance, user_profile=user.userprofile).first()
-            if not enrollment:
-                raise PermissionDenied()
+            # PS: is the default get exception ok, or should we raise PermissionDenied?
+            enrollment = Enrollment.objects.get(course_instance=course_instance, user_profile=user.userprofile)
```

Comment:
I would first guess that the exception directly affects the HTTP response status code. PermissionDenied leads to 403 Forbidden and other errors to 500 Internal Server Error. If this is the case, then we would want to keep 403 here.

Reply:
The `DoesNotExist` should yield 404, so it's choice between 403 and 404. I don't remember reasoning why 403 is used here. I tried to check git history, but there were no details. PR could include more details.

On a side note, I'm not sure how common `PS` is as a tag. I have used `FIXME` for things that shouldn't get merged/pushed, `TODO` for things that _should_ have an issue (or they have). `NOTE` to highlight things that are not apparent from the source and to make the comment more important than just comment without any tag. In this respect, I would write `FIXME` or `TODO` here to indicate that this part requries rework before merging. And this is just my personal opinion.

Reply:
Oh yes of course, if an object does not exist, the error must be HTTP 404 Not Found. Using 404 in this case could make sense since the problem really is that the enrollment is not found. This is typical for teachers and LTI services are the only reason why teachers need to enroll at the moment. The course points page does not separate teachers from students if everyone is enrolled, but that is another issue. Therefore, we could use the 404 response here unless we find a reason why PermissionDenied 403 has been used in the past.

Adding roles to enrollments so that teachers could always be enrolled:
https://github.com/apluslms/a-plus/issues/602

Reply:
So if I leave it as is, raising DoesNotExist would cause HTTP 404 by the framework? Or should I do something else to cause 404 response?

Reply:
I think we shall switch to the 404 error in this case. It would be best if the user can still get a clear error message instead of a very generic "404 not found". The message should say that the enrollment is missing and is required to access the LTI service. A+ already has custom error messages in at least 403 Forbidden pages, so I guess it is easy to do it in the 404 page too. Maybe it was done with the `messages` module, but I don't remember exactly.

This is one example of a customized 404 error message, but it is defined in a Permission class and this LTI code does not use permissions.
https://github.com/apluslms/a-plus/blob/24bd64eae94923a95729a8793bdf6a8babe49b61/course/permissions.py#L57

Django has the shortcut `get_object_or_404` that could be used assuming that you can modify the error message with it. At any rate, there is also the exception `Http404` that can probably be raised with a message string. Note that this message to the user shall be translated to Finnish too. (There is the function `_("hello message")` for translated strings.) Our A+ orientation wiki and the a-plus/doc/README.md also explain how to add translations (to the django.po file).

https://docs.djangoproject.com/en/2.2/topics/http/shortcuts/#get-object-or-404
https://docs.djangoproject.com/en/2.2/topics/i18n/translation/